frontend/src/components/widgets/FileUploader/FileUploader.tsx

```diff
@@ -0,0 +1,163 @@
+  width: number
+}
+
+interface State {
+  status: "READY" | "READING" | "UPLOADING"
+  errorMessage?: string
+}
+
+class FileUploader extends React.PureComponent<Props, State> {
+  public constructor(props: Props) {
+    super(props)
+    this.state = {
+      status: "READY",
+    }
+  }
+
+  private handleFileRead = (
+    ev: ProgressEvent<FileReader>,
+    file: File
+  ): void => {
+    if (ev.target !== null) {
+      if (ev.target.result instanceof ArrayBuffer) {
+        this.props.widgetStateManager.sendUploadFileMessage(
+          this.props.element.get("id"),
+          file.name,
+          file.lastModified,
+          new Uint8Array(ev.target.result)
+        )
+      } else {
+        console.warn("This file is not ArrayBuffer type.")
+      }
+    }
+    this.setState({ status: "UPLOADING" })
+  }
+
+  private dropHandler = (
+    acceptedFiles: File[],
+    rejectedFiles: File[],
+    event: React.SyntheticEvent<HTMLElement>
+  ): void => {
+    const { element } = this.props
+    const maxSizeMb = element.get("maxUploadSizeMb")
+
+    if (rejectedFiles.length > 0) {
+      this.setState({
+        status: "READY",
+        errorMessage: `${rejectedFiles[0].type} files are not allowed`,
+      })
+      return
+    }
+
+    this.setState({ status: "READING" })
+    acceptedFiles.forEach((file: File) => {
+      const fileSizeMB = file.size / 1024 / 1024
+      if (fileSizeMB < maxSizeMb) {
+        const fileReader = new FileReader()
+        fileReader.onloadend = (ev: ProgressEvent<FileReader>) =>
+          this.handleFileRead(ev, file)
+        fileReader.readAsArrayBuffer(file)
+      } else {
+        this.setState({
+          status: "READY",
+          errorMessage: `The max file size allowed is ${maxSizeMb}MB`,
+        })
+      }
+    })
+  }
+
+  public componentDidUpdate(oldProps: Props): void {
+    const progress = this.props.element.get("progress")
+    const oldProgress = oldProps.element.get("progress")
+    if (
+      oldProgress !== 1 &&
+      progress === 1 &&
+      this.state.status === "UPLOADING"
+    ) {
+      this.setState({ status: "READY" })
+    }
+  }
+
+  closeErrorMessage = (): void => {
+    this.setState({ status: "READY", errorMessage: undefined })
+  }
+
+  renderErrorMessage = (): React.ReactNode => {
+    const { errorMessage } = this.state
+    return (
+      <div className="file-uploader-error">
+        <span className="file-uploader-error__text">{errorMessage}</span>
+        <Button
+          className="file-uploader-error__button"
+          outline
+          onClick={this.closeErrorMessage}
+        >
+          Ok
+        </Button>
+      </div>
+    )
+  }
+
+  public render = (): React.ReactNode => {
+    const { status, errorMessage } = this.state
+    const { element } = this.props
+    const accept: string[] = element.get("type").toArray()
+    const label: string = element.get("label")
+    return (
+      <div className="file-uploader">
```

Comment:
The `className` should be `Widget stFileUploader`.

Without `Widget`, the `<label>` doesn't look right.

And all Streamlit elements need to be wrapped in a DOM element with class `stElementName` (see our [style guide](https://github.com/streamlit/streamlit/wiki/Style-Guide#css--scss))